File: functional-tests/functional/android/src/test/java/com/here/android/test/ArraysTest.java
/*
 * Copyright (C) 2016-2019 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */
package com.here.android.test;

import static junit.framework.Assert.assertEquals;
import static junit.framework.Assert.assertNotNull;
import static junit.framework.Assert.assertTrue;

import android.os.Build;
import com.here.android.RobolectricApplication;
import java.util.Collections;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.annotation.Config;

@RunWith(RobolectricTestRunner.class)
@Config(sdk = Build.VERSION_CODES.M, application = RobolectricApplication.class)
public class ArraysTest {
  private static final String STRING_LIST_ITEM_1 = "Item1";
  private static final String STRING_LIST_ITEM_2 = "Item2";
  private static final String STRING_LIST_ITEM_3 = "Item3";
  private static final float FLOAT_VALUE_1 = 42.0f;
  private static final float FLOAT_VALUE_2 = 43.0f;
  private static final double DOUBLE_VALUE_1 = 42.0;
  private static final double DOUBLE_VALUE_2 = 43.0;
  private static final double DOUBLE_VALUE_3 = 44.0;
  private static final double DOUBLE_VALUE_4 = 45.0;
  private static final long LONG_VALUE_1 = 44L;
  private static final long LONG_VALUE_2 = 45L;
  private static final long LONG_VALUE_3 = 46L;
  private static final long LONG_VALUE_4 = 47L;
  private static final Byte BYTE_ITEM_1 = (byte) 4;
  private static final Byte BYTE_ITEM_2 = (byte) 5;
  private static final Short SHORT_ITEM_1 = (short) 4;
  private static final Short SHORT_ITEM_2 = (short) 5;
  private static final Short SHORT_ITEM_3 = (short) 6;
  private static final Short SHORT_ITEM_4 = (short) 7;
  private static final Integer INT_ITEM_1 = 6;
  private static final Integer INT_ITEM_2 = 7;

  @Test
  public void reverseStringArray() {
    List<String> stringList = java.util.Arrays.asList(STRING_LIST_ITEM_1, STRING_LIST_ITEM_2);

    List<String> resultsList = Arrays.reverseStringArray(stringList);

    assertNotNull(resultsList);
    assertEquals(2, resultsList.size());
    assertEquals(STRING_LIST_ITEM_2, resultsList.get(0));
    assertEquals(STRING_LIST_ITEM_1, resultsList.get(1));
  }

  @Test
  public void reverseStringArray_useLinkedList() {
    List<String> stringList = new LinkedList<>();
    stringList.add(STRING_LIST_ITEM_1);
    stringList.add(STRING_LIST_ITEM_2);

    List<String> resultsList = Arrays.reverseStringArray(stringList);

    assertNotNull(resultsList);
    assertEquals(2, resultsList.size());
    assertEquals(STRING_LIST_ITEM_2, resultsList.get(0));
    assertEquals(STRING_LIST_ITEM_1, resultsList.get(1));
  }

  @Test
  public void reverseStringArray_useNull() {
    List<String> resultsList = Arrays.reverseStringArray(null);

    assertNotNull(resultsList);
    assertEquals(0, resultsList.size());
  }

  @Test
  public void methodWithInlineArray() {
    List<Short> shortList = java.util.Arrays.asList(SHORT_ITEM_1, SHORT_ITEM_2);

    List<Short> resultsList = Arrays.reverseArrayInline(shortList);

    assertNotNull(resultsList);
    assertEquals(2, resultsList.size());
    assertEquals(SHORT_ITEM_2, resultsList.get(0));
    assertEquals(SHORT_ITEM_1, resultsList.get(1));
  }

  @Test
  public void reverseInt8Array() {
    List<Byte> byteList = java.util.Arrays.asList(BYTE_ITEM_1, BYTE_ITEM_2);

    List<Byte> resultsList = Arrays.reverseInt8Array(byteList);

    assertNotNull(resultsList);
    assertEquals(2, resultsList.size());
    assertEquals(BYTE_ITEM_2, resultsList.get(0));
    assertEquals(BYTE_ITEM_1, resultsList.get(1));
  }

  @Test
  public void reverseInt16Array() {
    List<Short> shortList = java.util.Arrays.asList(SHORT_ITEM_1, SHORT_ITEM_2);

    List<Short> resultsList = Arrays.reverseInt16Array(shortList);

    assertNotNull(resultsList);
    assertEquals(2, resultsList.size());
    assertEquals(SHORT_ITEM_2, resultsList.get(0));
    assertEquals(SHORT_ITEM_1, resultsList.get(1));
  }

  @Test
  public void reverseInt32Array() {
    List<Integer> intList = java.util.Arrays.asList(INT_ITEM_1, INT_ITEM_2);

    List<Integer> resultsList = Arrays.reverseInt32Array(intList);

    assertNotNull(resultsList);
    assertEquals(2, resultsList.size());
    assertEquals(INT_ITEM_2, resultsList.get(0));
    assertEquals(INT_ITEM_1, resultsList.get(1));
  }

  @Test
  public void reverseInt64Array() {
    List<Long> longList = java.util.Arrays.asList(LONG_VALUE_1, LONG_VALUE_2);

    List<Long> resultsList = Arrays.reverseInt64Array(longList);

    assertNotNull(resultsList);
    assertEquals(2, resultsList.size());
    assertEquals(LONG_VALUE_2, resultsList.get(0).longValue());
    assertEquals(LONG_VALUE_1, resultsList.get(1).longValue());
  }

  @Test
  public void reverseUint8Array() {
    List<Short> shortList = java.util.Arrays.asList(SHORT_ITEM_1, SHORT_ITEM_2);

    List<Short> resultsList = Arrays.reverseUint8Array(shortList);

    assertNotNull(resultsList);
    assertEquals(2, resultsList.size());
    assertEquals(SHORT_ITEM_2, resultsList.get(0));
    assertEquals(SHORT_ITEM_1, resultsList.get(1));
  }

  @Test
  public void reverseUint16Array() {
    List<Integer> longList = java.util.Arrays.asList(INT_ITEM_1, INT_ITEM_2);

    List<Integer> resultsList = Arrays.reverseUint16Array(longList);

    assertNotNull(resultsList);
    assertEquals(2, resultsList.size());
    assertEquals(INT_ITEM_2, resultsList.get(0));
    assertEquals(INT_ITEM_1, resultsList.get(1));
  }

  @Test
  public void reverseUint32Array() {
    List<Long> longList = java.util.Arrays.asList(LONG_VALUE_1, LONG_VALUE_2);

    List<Long> resultsList = Arrays.reverseUint32Array(longList);

    assertNotNull(resultsList);
    assertEquals(2, resultsList.size());
    assertEquals(LONG_VALUE_2, resultsList.get(0).longValue());
    assertEquals(LONG_VALUE_1, resultsList.get(1).longValue());
  }

  @Test
  public void reverseUint64Array() {
    List<Long> longList = java.util.Arrays.asList(LONG_VALUE_1, LONG_VALUE_2);

    List<Long> resultsList = Arrays.reverseUint64Array(longList);

    assertNotNull(resultsList);
    assertEquals(2, resultsList.size());
    assertEquals(LONG_VALUE_2, resultsList.get(0).longValue());
    assertEquals(LONG_VALUE_1, resultsList.get(1).longValue());
  }

  @Test
  public void reverseFloatArray() {
    List<Float> floatList = java.util.Arrays.asList(FLOAT_VALUE_1, FLOAT_VALUE_2);

    List<Float> resultsList = Arrays.reverseFloatArray(floatList);

    assertNotNull(resultsList);
    assertEquals(2, resultsList.size());
    assertEquals(FLOAT_VALUE_2, resultsList.get(0));
    assertEquals(FLOAT_VALUE_1, resultsList.get(1));
  }

  @Test
  public void reverseDoubleArray() {
    List<Double> doubleList = java.util.Arrays.asList(DOUBLE_VALUE_1, DOUBLE_VALUE_2);

    List<Double> resultsList = Arrays.reverseDoubleArray(doubleList);

    assertNotNull(resultsList);
    assertEquals(2, resultsList.size());
    assertEquals(DOUBLE_VALUE_2, resultsList.get(0));
    assertEquals(DOUBLE_VALUE_1, resultsList.get(1));
  }

  @Test
  public void methodWithBooleanArray() {
    List<Boolean> booleanList = java.util.Arrays.asList(true, true, false);

    List<Boolean> resultsList = Arrays.reverseBoolArray(booleanList);

    assertNotNull(resultsList);
    assertEquals(3, resultsList.size());
    assertEquals(false, resultsList.get(0).booleanValue());
    assertEquals(true, resultsList.get(1).booleanValue());
    assertEquals(true, resultsList.get(2).booleanValue());
  }

  @Test
  public void reverseStructArray() {
    Arrays.BasicStruct struct1 = new Arrays.BasicStruct();
    Arrays.BasicStruct struct2 = new Arrays.BasicStruct();
    struct1.value = DOUBLE_VALUE_1;
    struct2.value = DOUBLE_VALUE_2;
    List<Arrays.BasicStruct> structList = java.util.Arrays.asList(struct1, struct2);

    List<Arrays.BasicStruct> resultsList = Arrays.reverseStructArray(structList);

    assertNotNull(resultsList);
    assertEquals(2, resultsList.size());
    assertEquals(DOUBLE_VALUE_2, resultsList.get(0).value);
    assertEquals(DOUBLE_VALUE_1, resultsList.get(1).value);
  }

  @Test
  public void reverseMapsArray() {
    Map<Short, String> map0 = Collections.emptyMap();
    Map<Short, String> map1 = new HashMap<>();
    map1.put((short) 42, STRING_LIST_ITEM_1);

    List<Map<Short, String>> input = java.util.Arrays.asList(map0, map1);

    List<Map<Short, String>> result = Arrays.reverseMapsArray(input);
    assertNotNull(result);
    assertEquals(2, result.size());
    assertEquals(map1, result.get(0));
    assertEquals(map0, result.get(1));
  }

  @Test
  public void reverseArrayMapsArray() {
    Map<Short, List<String>> map0 = Collections.emptyMap();
    Map<Short, List<String>> map1 = new HashMap<>();
    map1.put((short) 42, Collections.singletonList(STRING_LIST_ITEM_1));

    List<Map<Short, List<String>>> input = java.util.Arrays.asList(map0, map1);

    List<Map<Short, List<String>>> result = Arrays.reverseArrayMapsArray(input);
    assertNotNull(result);
    assertEquals(2, result.size());
    assertEquals(map1, result.get(0));
    assertEquals(map0, result.get(1));
  }

  @Test
  public void methodWithExplicitInstanceArray() {
    SimpleInstantiableOne instance1 = InstancesFactory.createSimpleInstantiableOne();
    SimpleInstantiableOne instance2 = InstancesFactory.createSimpleInstantiableOne();
    instance1.setStringValue(STRING_LIST_ITEM_1);
    instance2.setStringValue(STRING_LIST_ITEM_2);
    List<SimpleInstantiableOne> instancesArray = java.util.Arrays.asList(instance1, instance2);

    List<SimpleInstantiableOne> resultsList = Arrays.reverseExplicitInstancesArray(instancesArray);

    assertNotNull(resultsList);
    assertEquals(2, resultsList.size());
    assertEquals(STRING_LIST_ITEM_2, resultsList.get(0).getStringValue());
    assertEquals(STRING_LIST_ITEM_1, resultsList.get(1).getStringValue());
  }

  @Test
  public void methodWithImplicitInstanceArray() {
    SimpleInstantiableOne instance1 = InstancesFactory.createSimpleInstantiableOne();
    SimpleInstantiableOne instance2 = InstancesFactory.createSimpleInstantiableOne();
    instance1.setStringValue(STRING_LIST_ITEM_1);
    instance2.setStringValue(STRING_LIST_ITEM_2);
    List<SimpleInstantiableOne> instancesArray = java.util.Arrays.asList(instance1, instance2);

    List<SimpleInstantiableOne> resultsList = Arrays.reverseImplicitInstancesArray(instancesArray);

    assertNotNull(resultsList);
    assertEquals(2, resultsList.size());
    assertEquals(STRING_LIST_ITEM_2, resultsList.get(0).getStringValue());
    assertEquals(STRING_LIST_ITEM_1, resultsList.get(1).getStringValue());
  }

  @Test
  public void reverseNestedPrimitiveArray() {
    List<List<Double>> nestedDoubleList =
        java.util.Arrays.asList(
            java.util.Arrays.asList(DOUBLE_VALUE_1, DOUBLE_VALUE_2),
            java.util.Arrays.asList(DOUBLE_VALUE_3, DOUBLE_VALUE_4));

    List<List<Double>> resultsList = Arrays.reverseNestedPrimitiveArray(nestedDoubleList);

    assertNotNull(resultsList);
    assertEquals(2, resultsList.size());
    List<Double> firstResult = resultsList.get(0);
    List<Double> secondResult = resultsList.get(1);
    assertEquals(2, firstResult.size());
    assertEquals(2, secondResult.size());
    assertEquals(DOUBLE_VALUE_4, firstResult.get(0));
    assertEquals(DOUBLE_VALUE_3, firstResult.get(1));
    assertEquals(DOUBLE_VALUE_2, secondResult.get(0));
    assertEquals(DOUBLE_VALUE_1, secondResult.get(1));
  }

  @Test
  public void reverseNestedStructArray() {
    Arrays.BasicStruct struct1 = new Arrays.BasicStruct();
    Arrays.BasicStruct struct2 = new Arrays.BasicStruct();
    Arrays.BasicStruct struct3 = new Arrays.BasicStruct();
    Arrays.BasicStruct struct4 = new Arrays.BasicStruct();
    struct1.value = DOUBLE_VALUE_1;
    struct2.value = DOUBLE_VALUE_2;
    struct3.value = DOUBLE_VALUE_3;
    struct4.value = DOUBLE_VALUE_4;
    List<List<Arrays.BasicStruct>> nestedStructList =
        java.util.Arrays.asList(
            java.util.Arrays.asList(struct1, struct2), java.util.Arrays.asList(struct3, struct4));

    List<List<Arrays.BasicStruct>> resultsList = Arrays.reverseNestedStructArray(nestedStructList);

    assertNotNull(resultsList);
    assertEquals(2, resultsList.size());
    List<Arrays.BasicStruct> firstResult = resultsList.get(0);
    List<Arrays.BasicStruct> secondResult = resultsList.get(1);
    assertEquals(2, firstResult.size());
    assertEquals(2, secondResult.size());
    assertEquals(DOUBLE_VALUE_4, firstResult.get(0).value);
    assertEquals(DOUBLE_VALUE_3, firstResult.get(1).value);
    assertEquals(DOUBLE_VALUE_2, secondResult.get(0).value);
    assertEquals(DOUBLE_VALUE_1, secondResult.get(1).value);
  }

  @Test
  public void reverseNestedArraysInline() {
    List<List<Long>> nestedLongList =
        java.util.Arrays.asList(
            java.util.Arrays.asList(LONG_VALUE_1, LONG_VALUE_2),
            java.util.Arrays.asList(LONG_VALUE_3, LONG_VALUE_4));

    List<List<Long>> resultsList = Arrays.reverseNestedArraysInline(nestedLongList);

    assertNotNull(resultsList);
    assertEquals(2, resultsList.size());
    List<Long> firstResult = resultsList.get(0);
    List<Long> secondResult = resultsList.get(1);
    assertEquals(2, firstResult.size());
    assertEquals(2, secondResult.size());
    assertEquals(LONG_VALUE_4, firstResult.get(0).longValue());
    assertEquals(LONG_VALUE_3, firstResult.get(1).longValue());
    assertEquals(LONG_VALUE_2, secondResult.get(0).longValue());
    assertEquals(LONG_VALUE_1, secondResult.get(1).longValue());
  }

  @Test
  public void mergeArraysOfStructsWithArrays() {
    SimpleInstantiableOne instance = InstancesFactory.createSimpleInstantiableOne();
    SimpleInstantiableOne otherInstance = InstancesFactory.createSimpleInstantiableOne();
    instance.setStringValue(STRING_LIST_ITEM_1);
    otherInstance.setStringValue(STRING_LIST_ITEM_2);
    List<SimpleInstantiableOne> instancesArray1 = java.util.Arrays.asList(instance, otherInstance);
    SimpleInstantiableOne instance2 = InstancesFactory.createSimpleInstantiableOne();
    instance2.setStringValue(STRING_LIST_ITEM_3);
    List<SimpleInstantiableOne> instancesArray2 = Collections.singletonList(instance2);
    Arrays.FancyStruct struct1 =
        new Arrays.FancyStruct(
            java.util.Arrays.asList("struct", "1"),
            java.util.Arrays.asList(SHORT_ITEM_1, SHORT_ITEM_2),
            instancesArray1);
    Arrays.FancyStruct struct2 =
        new Arrays.FancyStruct(
            java.util.Arrays.asList("struct", "2"),
            java.util.Arrays.asList(SHORT_ITEM_3, SHORT_ITEM_3),
            instancesArray2);
    Arrays.FancyStruct struct3 =
        new Arrays.FancyStruct(
            java.util.Arrays.asList("struct", "3"),
            java.util.Arrays.asList(SHORT_ITEM_4, SHORT_ITEM_4),
            null);
    List<Arrays.FancyStruct> fancyStructList1 = java.util.Arrays.asList(struct1, struct2);
    List<Arrays.FancyStruct> fancyStructList2 = Collections.singletonList(struct3);

    List<Arrays.FancyStruct> resultsList =
        Arrays.mergeArraysOfStructsWithArrays(fancyStructList1, fancyStructList2);

    assertEquals(3, resultsList.size());
    assertEquals("struct", resultsList.get(0).messages.get(0));
    assertEquals("3", resultsList.get(0).messages.get(1));
    assertTrue(resultsList.get(0).instances.isEmpty());
    assertTrue(resultsList.get(1).messages.get(1).contains("2"));
    assertEquals(1, resultsList.get(1).instances.size());
    assertEquals(STRING_LIST_ITEM_3, resultsList.get(1).instances.get(0).getStringValue());
    assertTrue(resultsList.get(2).messages.get(1).contains("1"));
    assertEquals(2, resultsList.get(2).instances.size());
    assertEquals(STRING_LIST_ITEM_1, resultsList.get(2).instances.get(0).getStringValue());
    assertEquals(STRING_LIST_ITEM_2, resultsList.get(2).instances.get(1).getStringValue());
  }

  @Test
  public void reverseArrayOfAliases_emptyList() {
    List<String> stringList = new LinkedList<>();

    List<String> resultsList = Arrays.reverseArrayOfAliases(stringList);

    assertNotNull(resultsList);
    assertEquals(0, resultsList.size());
  }

  @Test
  public void reverseArrayOfAliases_reversesArray() {
    List<String> stringList = java.util.Arrays.asList(STRING_LIST_ITEM_1, STRING_LIST_ITEM_2);

    List<String> resultsList = Arrays.reverseArrayOfAliases(stringList);

    assertNotNull(resultsList);
    assertEquals(2, resultsList.size());
    assertEquals(STRING_LIST_ITEM_2, resultsList.get(0));
    assertEquals(STRING_LIST_ITEM_1, resultsList.get(1));
  }
}

File: functional-tests/functional/android-kotlin/src/test/kotlin/com/here/android/test/ArraysTest.kt
/*
 * Copyright (C) 2016-2025 HERE Europe B.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 * License-Filename: LICENSE
 */
package com.here.android.test

import com.here.android.RobolectricApplication
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner
import org.robolectric.annotation.Config

@RunWith(RobolectricTestRunner::class)
@Config(application = RobolectricApplication::class)
class ArraysTest {
    val floatDelta = 0.000000000001f
    val doubleDelta = 0.000000000001

    @org.junit.Test
    fun reverseStringArray() {
        val first = "abc"
        val second = "def"

        val stringList = listOf(first, second)
        val resultsList = Arrays.reverseStringArray(stringList)

        assertEquals(2, resultsList.size)
        assertEquals("def", resultsList[0])
        assertEquals("abc", resultsList[1])
    }

    @org.junit.Test
    fun methodWithInlineArray() {
        val first: Short = 77
        val second: Short = 21

        val shortList = listOf(first, second)
        val resultsList = Arrays.reverseArrayInline(shortList)

        assertEquals(2, resultsList.size)
        assertEquals(second, resultsList[0])
        assertEquals(first, resultsList[1])
    }

    @org.junit.Test
    fun reverseInt8Array() {
        val first: Byte = 55
        val second: Byte = 33

        val byteList = listOf(first, second)
        val resultsList = Arrays.reverseInt8Array(byteList)

        assertEquals(2, resultsList.size)
        assertEquals(second, resultsList[0])
        assertEquals(first, resultsList[1])
    }

    @org.junit.Test
    fun reverseInt16Array() {
        val first: Short = 77
        val second: Short = 21

        val shortList = listOf(first, second)
        val resultsList = Arrays.reverseInt16Array(shortList)

        assertEquals(2, resultsList.size)
        assertEquals(second, resultsList[0])
        assertEquals(first, resultsList[1])
    }

    @org.junit.Test
    fun reverseInt32Array() {
        val first: Int = 57
        val second: Int = 23

        val intList = listOf(first, second)
        val resultsList = Arrays.reverseInt32Array(intList)

        assertEquals(2, resultsList.size)
        assertEquals(second, resultsList[0])
        assertEquals(first, resultsList[1])
    }

    @org.junit.Test
    fun reverseInt64Array() {
        val first: Long = 17L
        val second: Long = 23L

        val longList = listOf(first, second)
        val resultsList = Arrays.reverseInt64Array(longList)

        assertEquals(2, resultsList.size)
        assertEquals(second, resultsList[0])
        assertEquals(first, resultsList[1])
    }

    @org.junit.Test
    fun reverseUint8Array() {
        val first: Short = 77
        val second: Short = 21

        val shortList = listOf(first, second)
        val resultsList = Arrays.reverseUint8Array(shortList)

        assertEquals(2, resultsList.size)
        assertEquals(second, resultsList[0])
        assertEquals(first, resultsList[1])
    }

    @org.junit.Test
    fun reverseUint16Array() {
        val first: Int = 57
        val second: Int = 23

        val intList = listOf(first, second)
        val resultsList = Arrays.reverseUint16Array(intList)

        assertEquals(2, resultsList.size)
        assertEquals(second, resultsList[0])
        assertEquals(first, resultsList[1])
    }

    @org.junit.Test
    fun reverseUint32Array() {
        val first: Long = 17L
        val second: Long = 23L

        val longList = listOf(first, second)
        val resultsList = Arrays.reverseUint32Array(longList)

        assertEquals(2, resultsList.size)
        assertEquals(second, resultsList[0])
        assertEquals(first, resultsList[1])
    }

    @org.junit.Test
    fun reverseUint64Array() {
        val first: Long = 17L
        val second: Long = 23L

        val longList = listOf(first, second)
        val resultsList = Arrays.reverseUint64Array(longList)

        assertEquals(2, resultsList.size)
        assertEquals(second, resultsList[0])
        assertEquals(first, resultsList[1])
    }

    @org.junit.Test
    fun reverseFloatArray() {
        val first: Float = 77.77f
        val second: Float = 23.23f

        val floatList = listOf(first, second)
        val resultsList = Arrays.reverseFloatArray(floatList)

        assertEquals(2, resultsList.size)
        assertEquals(second, resultsList[0], floatDelta)
        assertEquals(first, resultsList[1], floatDelta)
    }

    @org.junit.Test
    fun reverseDoubleArray() {
        val first: Double = 22.22
        val second: Double = 66.66

        val doubleList = listOf(first, second)
        val resultsList = Arrays.reverseDoubleArray(doubleList)

        assertEquals(2, resultsList.size)
        assertEquals(second, resultsList[0], doubleDelta)
        assertEquals(first, resultsList[1], doubleDelta)
    }

    @org.junit.Test
    fun methodWithBooleanArray() {
        val booleanList = listOf<Boolean>(true, true, false)
        val resultsList = Arrays.reverseBoolArray(booleanList)

        assertEquals(3, resultsList.size)
        assertFalse(resultsList[0])
        assertTrue(resultsList[1])
        assertTrue(resultsList[2])
    }

    @org.junit.Test
    fun reverseStructArray() {
        val first = Arrays.BasicStruct(22.22)
        val second = Arrays.BasicStruct(33.33)

        val structList = listOf(first, second)
        val resultsList = Arrays.reverseStructArray(structList)

        assertEquals(2, resultsList.size)
        assertEquals(second.value, resultsList[0].value, doubleDelta)
        assertEquals(first.value, resultsList[1].value, doubleDelta)
    }

    @org.junit.Test
    fun reverseMapsArray() {
        val first = mapOf(32.toShort() to "abc")
        val second = mapOf<Short, String>()

        val input = listOf(first, second)
        val result = Arrays.reverseMapsArray(input)

        assertEquals(2, result.size)
        assertEquals(second, result[0])
        assertEquals(first, result[1])
    }

    @org.junit.Test
    fun reverseArrayMapsArray() {
        val first = mapOf(42.toShort() to listOf("john", "doe"))
        val second = mapOf(44.toShort() to listOf("example", "dot", "com"))

        val input = listOf(first, second)
        val result = Arrays.reverseArrayMapsArray(input)

        assertEquals(2, result.size)
        assertEquals(second, result[0])
        assertEquals(first, result[1])
    }

    @org.junit.Test
    fun methodWithExplicitInstanceArray() {
        val first = InstancesFactory.createSimpleInstantiableOne()
        first.setStringValue("FIRST STRING")

        val second = InstancesFactory.createSimpleInstantiableOne()
        second.setStringValue("SECOND STRING")

        val instancesArray = listOf(first, second)
        val resultsList = Arrays.reverseExplicitInstancesArray(instancesArray)

        assertEquals(2, resultsList.size)
        assertEquals(second.getStringValue(), resultsList[0].getStringValue())
        assertEquals(first.getStringValue(), resultsList[1].getStringValue())
    }

    @org.junit.Test
    fun methodWithImplicitInstanceArray() {
        val first = InstancesFactory.createSimpleInstantiableOne()
        first.setStringValue("FIRST STRING")

        val second = InstancesFactory.createSimpleInstantiableOne()
        second.setStringValue("SECOND STRING")

        val instancesArray = listOf(first, second)
        val resultsList: List<SimpleInstantiableOne> = Arrays.reverseImplicitInstancesArray(instancesArray)

        assertEquals(2, resultsList.size)
        assertEquals(second.getStringValue(), resultsList[0].getStringValue())
        assertEquals(first.getStringValue(), resultsList[1].getStringValue())
    }

    @org.junit.Test
    fun reverseNestedPrimitiveArray() {
        val nestedDoubleList = listOf(listOf(11.11, 22.22), listOf(33.33, 44.44))
        val resultsList = Arrays.reverseNestedPrimitiveArray(nestedDoubleList)

        assertEquals(2, resultsList.size)

        val firstResult = resultsList[0]
        assertEquals(2, firstResult.size)
        assertEquals(nestedDoubleList[1][1], firstResult[0], doubleDelta)
        assertEquals(nestedDoubleList[1][0], firstResult[1], doubleDelta)

        val secondResult = resultsList[1]
        assertEquals(2, secondResult.size)
        assertEquals(nestedDoubleList[0][1], secondResult[0], doubleDelta)
        assertEquals(nestedDoubleList[0][0], secondResult[1], doubleDelta)
    }

    @org.junit.Test
    fun reverseNestedStructArray() {
        val nestedStructList = listOf(
            listOf(Arrays.BasicStruct(11.11), Arrays.BasicStruct(22.22)),
            listOf(Arrays.BasicStruct(33.33), Arrays.BasicStruct(44.44))
        )

        val resultsList = Arrays.reverseNestedStructArray(nestedStructList)

        val firstResult = resultsList[0]
        assertEquals(2, firstResult.size)
        assertEquals(nestedStructList[1][1].value, firstResult[0].value, doubleDelta)
        assertEquals(nestedStructList[1][0].value, firstResult[1].value, doubleDelta)

        val secondResult = resultsList[1]
        assertEquals(2, secondResult.size)
        assertEquals(nestedStructList[0][1].value, secondResult[0].value, doubleDelta)
        assertEquals(nestedStructList[0][0].value, secondResult[1].value, doubleDelta)
    }

    @org.junit.Test
    fun reverseNestedArraysInline() {
        val nestedLongList = listOf(listOf(21L, 31L), listOf(45L, 75L))
        val resultsList = Arrays.reverseNestedArraysInline(nestedLongList)

        assertEquals(2, resultsList.size)

        val firstResult = resultsList[0]
        assertEquals(2, firstResult.size)
        assertEquals(nestedLongList[1][1], firstResult[0])
        assertEquals(nestedLongList[1][0], firstResult[1])

        val secondResult = resultsList[1]
        assertEquals(2, secondResult.size)
        assertEquals(nestedLongList[0][1], secondResult[0])
        assertEquals(nestedLongList[0][0], secondResult[1])
    }

    @org.junit.Test
    fun mergeArraysOfStructsWithArrays() {
        val instance1 = InstancesFactory.createSimpleInstantiableOne()
        instance1.setStringValue("Instance 1: STRING")

        val instance2 = InstancesFactory.createSimpleInstantiableOne()
        instance2.setStringValue("Instance 2: STRING")

        val instance3 = InstancesFactory.createSimpleInstantiableOne()
        instance3.setStringValue("Instance 3: STRING")

        val instancesArray1 = listOf(instance1, instance2)
        val instancesArray2 = listOf(instance3)
        val instancesArray3 = listOf<SimpleInstantiableOne>()

        val struct1 = Arrays.FancyStruct(
            listOf("struct", "1"),
            listOf(1.toShort(), 2.toShort()),
            instancesArray1
        )

        val struct2 = Arrays.FancyStruct(
            listOf("struct", "2"),
            listOf(3.toShort(), 4.toShort()),
            instancesArray2
        )

        val struct3 =  Arrays.FancyStruct(
            listOf("struct", "3"),
            listOf(5.toShort(), 6.toShort()),
            instancesArray3
        )

        val fancyStructList1 = listOf(struct1, struct2)
        val fancyStructList2 = listOf(struct3)
        val resultsList = Arrays.mergeArraysOfStructsWithArrays(fancyStructList1, fancyStructList2)

        assertEquals(3, resultsList.size)

        assertEquals("struct", resultsList[0].messages[0])
        assertEquals("3", resultsList[0].messages[1])
        assertTrue(resultsList[0].instances.isEmpty())

        assertEquals("2", resultsList[1].messages[1])
        assertEquals(1, resultsList[1].instances.size)
        assertEquals("Instance 3: STRING", resultsList[1].instances[0].getStringValue())

        assertEquals("1", resultsList[2].messages[1])
        assertEquals(2, resultsList[2].instances.size)
        assertEquals("Instance 1: STRING", resultsList[2].instances[0].getStringValue())
        assertEquals("Instance 2: STRING", resultsList[2].instances[1].getStringValue())
    }

    @org.junit.Test
    fun reverseArrayOfAliases_emptyList() {
        val stringList = listOf<String>()
        val resultsList = Arrays.reverseArrayOfAliases(stringList)

        assertEquals(0, resultsList.size)
    }

    @org.junit.Test
    fun reverseArrayOfAliases_reversesArray() {
        val stringList = listOf("abc", "def")
        val resultsList = Arrays.reverseArrayOfAliases(stringList)

        assertEquals(2, resultsList.size)
        assertEquals("def", resultsList[0])
        assertEquals("abc", resultsList[1])
    }
}